integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/ContainerBasedKafkaCluster.java

```diff
@@ -0,0 +1,226 @@
+                .collect(Collectors.joining(","));
+    }
+
+    private Stream<GenericContainer<?>> allContainers() {
+        return Stream.concat(
+                this.brokers.stream(),
+                Stream.ofNullable(this.zookeeper));
+    }
+
+    @Override
+    @SneakyThrows
+    public void start() {
+        try {
+            if (zookeeper != null) {
+                zookeeper.start();
+                Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                    Container.ExecResult result = this.zookeeper.execInContainer(
+                            "sh", "-c",
+                            String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s whoami", ZOOKEEPER_PORT));
+                    LOGGER.debug("zookeeper running {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                    return result.getExitCode() == 0;
+                });
+            }
+            Startables.deepStart(brokers.stream()).get(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            stop();
+            throw new RuntimeException("startup failed or timed out", e);
+        }
+
+        if (clusterConfig.isKraftMode()) {
+            Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                Container.ExecResult result = this.brokers.iterator().next().execInContainer(
+                        "sh", "-c",
+                        "kafka-metadata-shell.sh --snapshot /bitnami/kafka/data/__cluster_metadata-0/*.log ls /brokers");
+                LOGGER.debug("kraft cluster ready probe {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                Optional<String> brokers = Optional.ofNullable(result.getStdout());
+                return brokers.map(b -> Arrays.stream(b.split(System.lineSeparator())).count()).map(count -> count.intValue() == this.clusterConfig.getBrokersNum())
+                        .orElse(false);
+            });
+        }
+        else {
+            Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                Container.ExecResult result = this.zookeeper.execInContainer(
+                        "sh", "-c",
+                        String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s ls /brokers/ids | tail -n 1",
+                                ZOOKEEPER_PORT));
+                LOGGER.debug("zookeeper cluster ready probe {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                String brokers = result.getStdout();
```

Comment:
Given where this is in the method I thought this was starting a traditional kafka cluster using Zookeeper rather than KRaft quorum controllers but then I don't see where it actually starts the brokers. 

This is just polling zookeeper, AFICT, to see when all the brokers have reported as started. 

What am I missing?

Reply:
I actually pushed this responsibility down the the native test containers.

### Zookeeper awaits until ZK is minimally available to a ZK client.

Kafka will actually fail to startup if ZK is not ready, so the intent of this is avoid this possibility.

https://github.com/k-wall/zookeeper-native/blob/main/zookeeper-server/src/main/java/io/kroxylicious/test/zookeeper/server/ZookeeperPoller.java

### Kafka awaits until the expected number of brokers are visible through the AdminAPI. 
Intent is avoid the possible that tests start before the cluster they specified is actually ready.

https://github.com/ozangunalp/kafka-native/pull/5